File: bindings/python/src/graphqlite/connection.py
"""GraphQLite connection wrapper for SQLite."""

import json
import os
import platform
import sqlite3
from pathlib import Path
from typing import Any, Iterator, Optional, Union


class CypherResult:
    """Result from a Cypher query, iterable as rows."""

    def __init__(self, data: list[dict[str, Any]], columns: list[str]):
        self._data = data
        self._columns = columns

    def __iter__(self) -> Iterator[dict[str, Any]]:
        return iter(self._data)

    def __len__(self) -> int:
        return len(self._data)

    def __getitem__(self, index: int) -> dict[str, Any]:
        return self._data[index]

    @property
    def columns(self) -> list[str]:
        """Column names from the query."""
        return self._columns

    def to_list(self) -> list[dict[str, Any]]:
        """Return results as a list of dictionaries."""
        return self._data


class Connection:
    """GraphQLite database connection with Cypher query support."""

    def __init__(self, conn: sqlite3.Connection, extension_path: Optional[str] = None):
        """
        Initialize GraphQLite connection.

        Args:
            conn: SQLite database connection
            extension_path: Path to graphqlite extension (auto-detected if None)
        """
        self._conn = conn
        self._load_extension(extension_path)

    def _find_extension(self) -> str:
        """Find the GraphQLite extension library."""
        system = platform.system()

        if system == "Darwin":
            ext_name = "graphqlite.dylib"
        elif system == "Linux":
            ext_name = "graphqlite.so"
        elif system == "Windows":
            ext_name = "graphqlite.dll"
        else:
            raise OSError(f"Unsupported platform: {system}")

        # Search paths in order of preference
        search_paths = [
            # Bundled with package
            Path(__file__).parent / ext_name,
            # Development build
            Path(__file__).parent.parent.parent.parent.parent / "build" / ext_name,
            # System-wide
            Path("/usr/local/lib") / ext_name,
            Path("/usr/lib") / ext_name,
        ]

        # Check GRAPHQLITE_EXTENSION_PATH environment variable
        env_path = os.environ.get("GRAPHQLITE_EXTENSION_PATH")
        if env_path:
            search_paths.insert(0, Path(env_path))

        for path in search_paths:
            if path.exists():
                return str(path.resolve())

        raise FileNotFoundError(
            f"GraphQLite extension not found. Searched: {[str(p) for p in search_paths]}\n"
            f"Set GRAPHQLITE_EXTENSION_PATH or build the extension with 'make extension'"
        )

    def _load_extension(self, extension_path: Optional[str] = None) -> None:
        """Load the GraphQLite SQLite extension."""
        if extension_path is None:
            extension_path = self._find_extension()

        # Enable extension loading
        try:
            self._conn.enable_load_extension(True)
        except AttributeError as e:
            raise RuntimeError(
                "SQLite extension loading not available. "
                "Your Python's sqlite3 module may not support extensions.\n"
                "On macOS with MacPorts/Homebrew, try:\n"
                "  DYLD_LIBRARY_PATH=/opt/local/lib python your_script.py"
            ) from e

        # Load extension (remove file extension for SQLite)
        ext_path = Path(extension_path)
        load_path = str(ext_path.parent / ext_path.stem)

        try:
            self._conn.load_extension(load_path)
        except sqlite3.OperationalError as e:
            error_msg = str(e).lower()
            if "not authorized" in error_msg:
                raise RuntimeError(
                    "SQLite extension loading is disabled. "
                    "The system SQLite may not allow extensions.\n"
                    "On macOS, try using Homebrew or MacPorts Python with:\n"
                    "  DYLD_LIBRARY_PATH=/opt/local/lib python your_script.py"
                ) from e
            raise

        # Verify extension loaded
        cursor = self._conn.execute("SELECT graphqlite_test()")
        result = cursor.fetchone()
        if not result or "successfully" not in result[0].lower():
            raise RuntimeError("Failed to initialize GraphQLite extension")

    def cypher(self, query: str, params: Optional[dict[str, Any]] = None) -> CypherResult:
        """
        Execute a Cypher query with optional parameters.

        Args:
            query: Cypher query string, may contain $param placeholders
            params: Optional dictionary of parameter values

        Returns:
            CypherResult object with query results

        Raises:
            sqlite3.Error: If the query fails

        Example:
            >>> db.cypher("MATCH (n) WHERE n.name = $name RETURN n", {"name": "Alice"})
        """
        try:
            if params:
                params_json = json.dumps(params)
                cursor = self._conn.execute("SELECT cypher(?, ?)", (query, params_json))
            else:
                cursor = self._conn.execute("SELECT cypher(?)", (query,))
        except sqlite3.Error as e:
            # Parse structured JSON error from extension
            err_str = str(e)
            try:
                err_data = json.loads(err_str)
                if isinstance(err_data, dict) and "error" in err_data:
                    raise sqlite3.Error(err_data["error"]) from None
            except (json.JSONDecodeError, TypeError):
                pass
            raise

        row = cursor.fetchone()

        if row is None or row[0] is None:
            return CypherResult([], [])

        result_str = row[0]

        # Parse JSON result
        try:
            data = json.loads(result_str)
        except json.JSONDecodeError:
            # Non-JSON result (scalar or legacy error)
            if result_str.startswith("Error") or result_str.startswith("{\"error\""):
                raise sqlite3.Error(result_str)
            return CypherResult([{"result": result_str}], ["result"])

        # Handle different result formats
        if isinstance(data, list):
            if len(data) == 0:
                return CypherResult([], [])
            if isinstance(data[0], dict):
                columns = list(data[0].keys()) if data else []
                return CypherResult(data, columns)
            # List of scalars - this happens when C returns raw JSON array
            # for single-cell queries (e.g., range(), tail(), graph algorithms)
            # Treat as single row with the original JSON string as value
            return CypherResult([{"result": result_str}], ["result"])
        elif isinstance(data, dict):
            return CypherResult([data], list(data.keys()))
        else:
            return CypherResult([{"result": data}], ["result"])

    def execute(self, sql: str, parameters: tuple = ()) -> sqlite3.Cursor:
        """Execute a raw SQL query."""
        return self._conn.execute(sql, parameters)

    def commit(self) -> None:
        """Commit the current transaction."""
        self._conn.commit()

    def rollback(self) -> None:
        """Rollback the current transaction."""
        self._conn.rollback()

    def close(self) -> None:
        """Close the database connection."""
        self._conn.close()

    def __enter__(self) -> "Connection":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()

    @property
    def sqlite_connection(self) -> sqlite3.Connection:
        """Access the underlying SQLite connection."""
        return self._conn


def connect(
    database: Union[str, Path] = ":memory:",
    extension_path: Optional[str] = None,
    **kwargs
) -> Connection:
    """
    Open a GraphQLite database connection.

    Args:
        database: Path to database file or ":memory:" for in-memory database
        extension_path: Path to graphqlite extension (auto-detected if None)
        **kwargs: Additional arguments passed to sqlite3.connect()

    Returns:
        Connection object with Cypher query support

    Example:
        >>> db = connect("graph.db")
        >>> db.cypher("CREATE (n:Person {name: 'Alice'})")
        >>> results = db.cypher("MATCH (n:Person) RETURN n.name")
        >>> for row in results:
        ...     print(row["n.name"])
    """
    conn = sqlite3.connect(str(database), **kwargs)
    return Connection(conn, extension_path)


def wrap(conn: sqlite3.Connection, extension_path: Optional[str] = None) -> Connection:
    """
    Wrap an existing SQLite connection with GraphQLite support.

    Args:
        conn: Existing SQLite connection
        extension_path: Path to graphqlite extension (auto-detected if None)

    Returns:
        Connection object with Cypher query support

    Example:
        >>> import sqlite3
        >>> conn = sqlite3.connect("graph.db")
        >>> db = wrap(conn)
        >>> db.cypher("MATCH (n) RETURN count(n)")
    """
    return Connection(conn, extension_path)

File: bindings/rust/src/connection.rs
//! GraphQLite connection wrapper.

use crate::{CypherResult, Error, Result};
use crate::query_builder::CypherQuery;

#[cfg(not(feature = "bundled-extension"))]
use std::path::PathBuf;
use std::path::Path;

/// A GraphQLite database connection.
///
/// Wraps a SQLite connection with the GraphQLite extension loaded,
/// providing Cypher query support.
pub struct Connection {
    conn: rusqlite::Connection,
}

impl Connection {
    /// Open a database at the given path.
    ///
    /// # Arguments
    ///
    /// * `path` - Path to database file, or ":memory:" for in-memory database
    ///
    /// # Example
    ///
    /// ```no_run
    /// use graphqlite::Connection;
    ///
    /// let conn = Connection::open(":memory:")?;
    /// # Ok::<(), graphqlite::Error>(())
    /// ```
    pub fn open<P: AsRef<Path>>(path: P) -> Result<Self> {
        let conn = rusqlite::Connection::open(path)?;
        Self::from_rusqlite(conn)
    }

    /// Open an in-memory database.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use graphqlite::Connection;
    ///
    /// let conn = Connection::open_in_memory()?;
    /// # Ok::<(), graphqlite::Error>(())
    /// ```
    pub fn open_in_memory() -> Result<Self> {
        let conn = rusqlite::Connection::open_in_memory()?;
        Self::from_rusqlite(conn)
    }

    /// Create a GraphQLite connection from an existing rusqlite Connection.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use graphqlite::Connection;
    /// use rusqlite;
    ///
    /// let sqlite_conn = rusqlite::Connection::open_in_memory()?;
    /// let conn = Connection::from_rusqlite(sqlite_conn)?;
    /// # Ok::<(), graphqlite::Error>(())
    /// ```
    #[cfg(feature = "bundled-extension")]
    pub fn from_rusqlite(conn: rusqlite::Connection) -> Result<Self> {
        // Load the bundled extension (extracts from embedded binary)
        crate::platform::load_bundled_extension(&conn)?;
        Ok(Connection { conn })
    }

    /// Create a GraphQLite connection from an existing rusqlite Connection.
    #[cfg(not(feature = "bundled-extension"))]
    pub fn from_rusqlite(conn: rusqlite::Connection) -> Result<Self> {
        let extension_path = find_extension()?;
        load_extension(&conn, &extension_path)?;
        Ok(Connection { conn })
    }

    /// Create a connection with a custom extension path.
    ///
    /// # Arguments
    ///
    /// * `path` - Path to database file
    /// * `extension_path` - Path to the GraphQLite extension (.dylib, .so, or .dll)
    ///
    /// Note: This method is only available when the `bundled-extension` feature is disabled.
    #[cfg(not(feature = "bundled-extension"))]
    pub fn open_with_extension<P: AsRef<Path>, E: AsRef<std::path::Path>>(
        path: P,
        extension_path: E,
    ) -> Result<Self> {
        let conn = rusqlite::Connection::open(path)?;
        load_extension(&conn, extension_path.as_ref())?;
        Ok(Connection { conn })
    }

    /// Execute a Cypher query.
    ///
    /// # Arguments
    ///
    /// * `query` - Cypher query string
    ///
    /// # Returns
    ///
    /// A `CypherResult` containing the query results.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use graphqlite::Connection;
    ///
    /// let conn = Connection::open_in_memory()?;
    /// conn.cypher("CREATE (n:Person {name: 'Alice'})")?;
    /// let results = conn.cypher("MATCH (n:Person) RETURN n.name")?;
    /// # Ok::<(), graphqlite::Error>(())
    /// ```
    pub fn cypher(&self, query: &str) -> Result<CypherResult> {
        let result: Option<String> = self
            .conn
            .query_row("SELECT cypher(?1)", [query], |row| row.get(0))?;

        match result {
            Some(json_str) => {
                // Check for error response (legacy prefix or structured JSON)
                if json_str.starts_with("Error") || json_str.starts_with("{\"error\"") {
                    return Err(parse_structured_error(&json_str));
                }
                CypherResult::from_json(&json_str)
            }
            None => Ok(CypherResult::empty()),
        }
    }

    /// Execute a Cypher query with named parameters.
    ///
    /// Parameters are passed as a JSON object and bound inside the extension,
    /// preventing injection and eliminating the need for manual escaping.
    ///
    /// # Arguments
    ///
    /// * `query` - Cypher query string with `$param` placeholders
    /// * `params` - Parameter values as a `serde_json::Value` (must be an object)
    ///
    /// # Example
    ///
    /// ```no_run
    /// use graphqlite::Connection;
    /// use serde_json::json;
    ///
    /// let conn = Connection::open_in_memory()?;
    /// conn.cypher("CREATE (n:Person {name: 'Alice', age: 30})")?;
    /// let results = conn.cypher_with_params(
    ///     "MATCH (n:Person) WHERE n.name = $name RETURN n.name, n.age",
    ///     &json!({"name": "Alice"})
    /// )?;
    /// # Ok::<(), graphqlite::Error>(())
    /// ```
    #[deprecated(since = "0.4.0", note = "Use cypher_builder() instead")]
    pub fn cypher_with_params(&self, query: &str, params: &serde_json::Value) -> Result<CypherResult> {
        self.execute_cypher_with_params(query, params)
    }

    /// Internal: execute a parameterized Cypher query.
    pub(crate) fn execute_cypher_with_params(&self, query: &str, params: &serde_json::Value) -> Result<CypherResult> {
        let params_json = serde_json::to_string(params)
            .map_err(|e| Error::Cypher(format!("Failed to serialize params: {}", e)))?;
        let result: Option<String> = self
            .conn
            .query_row("SELECT cypher(?1, ?2)", rusqlite::params![query, params_json], |row| row.get(0))?;

        match result {
            Some(json_str) => {
                if json_str.starts_with("Error") || json_str.starts_with("{\"error\"") {
                    return Err(parse_structured_error(&json_str));
                }
                CypherResult::from_json(&json_str)
            }
            None => Ok(CypherResult::empty()),
        }
    }

    /// Create a builder for a parameterized Cypher query.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use graphqlite::Connection;
    ///
    /// let conn = Connection::open_in_memory()?;
    /// conn.cypher("CREATE (n:Person {name: 'Alice', age: 30})")?;
    /// let results = conn.cypher_builder("MATCH (n:Person) WHERE n.name = $name RETURN n")
    ///     .param("name", "Alice")
    ///     .run()?;
    /// # Ok::<(), graphqlite::Error>(())
    /// ```
    pub fn cypher_builder<'a>(&'a self, query: &'a str) -> CypherQuery<'a> {
        CypherQuery::new(self, query)
    }

    /// Execute raw SQL.
    ///
    /// Useful for queries that don't use Cypher, like checking schema
    /// or using algorithm results with `json_each()`.
    pub fn execute(&self, sql: &str) -> Result<usize> {
        Ok(self.conn.execute(sql, [])?)
    }

    /// Access the underlying rusqlite connection.
    pub fn sqlite_connection(&self) -> &rusqlite::Connection {
        &self.conn
    }
}

/// Find the GraphQLite extension library.
#[cfg(not(feature = "bundled-extension"))]
fn find_extension() -> Result<PathBuf> {
    let ext_name = if cfg!(target_os = "macos") {
        "graphqlite.dylib"
    } else if cfg!(target_os = "windows") {
        "graphqlite.dll"
    } else {
        "graphqlite.so"
    };

    // Search paths in order of preference
    let search_paths: Vec<PathBuf> = vec![
        // Environment variable
        std::env::var("GRAPHQLITE_EXTENSION_PATH")
            .ok()
            .map(PathBuf::from)
            .unwrap_or_default(),
        // Current directory build
        PathBuf::from("build").join(ext_name),
        // Relative to crate root (for development)
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap_or(Path::new("."))
            .parent()
            .unwrap_or(Path::new("."))
            .join("build")
            .join(ext_name),
        // System paths
        PathBuf::from("/usr/local/lib").join(ext_name),
        PathBuf::from("/usr/lib").join(ext_name),
    ];

    for path in search_paths {
        if path.exists() {
            return Ok(path);
        }
    }

    Err(Error::ExtensionNotFound(format!(
        "Could not find {}. Build with 'make extension' or set GRAPHQLITE_EXTENSION_PATH",
        ext_name
    )))
}

/// Load the GraphQLite extension into a connection.
#[cfg(not(feature = "bundled-extension"))]
fn load_extension(conn: &rusqlite::Connection, path: &std::path::Path) -> Result<()> {
    // Remove the file extension for SQLite's load_extension
    let load_path = path.with_extension("");

    unsafe {
        conn.load_extension_enable()?;
        conn.load_extension(&load_path, None)?;
        conn.load_extension_disable()?;
    }

    // Verify the extension loaded
    let test: String = conn.query_row("SELECT graphqlite_test()", [], |row| row.get(0))?;
    if !test.to_lowercase().contains("successfully") {
        return Err(Error::ExtensionNotFound(
            "Extension loaded but verification failed".to_string(),
        ));
    }

    Ok(())
}

/// Parse a structured JSON error from the extension into an Error.
/// Handles both structured `{"error":"msg","code":"CODE"}` and legacy `"Error: msg"` formats.
fn parse_structured_error(s: &str) -> Error {
    if let Ok(v) = serde_json::from_str::<serde_json::Value>(s) {
        if let Some(msg) = v.get("error").and_then(|e| e.as_str()) {
            return Error::Cypher(msg.to_string());
        }
    }
    Error::Cypher(s.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[cfg(not(feature = "bundled-extension"))]
    fn get_test_extension_path() -> Option<std::path::PathBuf> {
        let paths = [
            std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"))
                .parent()
                .unwrap()
                .parent()
                .unwrap()
                .join("build/graphqlite.dylib"),
            std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"))
                .parent()
                .unwrap()
                .parent()
                .unwrap()
                .join("build/graphqlite.so"),
        ];

        paths.into_iter().find(|p| p.exists())
    }

    #[test]
    #[cfg(not(feature = "bundled-extension"))]
    fn test_find_extension() {
        // This test may skip if extension isn't built
        if get_test_extension_path().is_none() {
            return;
        }
        assert!(find_extension().is_ok());
    }

    #[test]
    #[cfg(feature = "bundled-extension")]
    fn test_bundled_connection() {
        // Test that bundled extension works
        let conn = Connection::open_in_memory();
        assert!(conn.is_ok(), "Failed to open connection: {:?}", conn.err());
    }
}
